File: leetcode/src/p002AddTwoNumbers/solution.py
"""
You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.

Example

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8
Explanation: 342 + 465 = 807.
"""
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def addTwoNumbers(self, l1, l2):
        head = None
        tag = 0
        while l1 is not None or l2 is not None or tag != 0:
            sum3 = 0
            if l1 is not None:
                sum3 += l1.val
                l1 = l1.next
            if l2 is not None:
                sum3 += l2.val
                l2 = l2.next
            sum3 += tag
            x = sum3 % 10
            tag = sum3 // 10
            if head is None:
                head = ListNode(x)
                ind = head
            else:
                ind.next = ListNode(x)
                ind = ind.next
        return head

File: leetcode/src/p002AddTwoNumbers/Solution.java
package p002AddTwoNumbers;

/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode(int x) { val = x; } }
 */

// Definition for singly-linked list.

class ListNode {
	int val;
	ListNode next;

	ListNode(int x) {
		val = x;
	}

	public ListNode(int[] a) {
		if (a.length > 0) {
			this.val = a[0];
			ListNode end = this;
			for (int i = 1; i < a.length; i++) {
				ListNode temp = new ListNode(a[i]);
				end.next = temp;
				end = end.next;
			}
		}
	}

	@Override
	public String toString() {
		String res = "[";
		ListNode temp = this;
		while (temp != null) {
			res += temp.val;
			res += " -> ";
			temp = temp.next;
		}
		res += "null";
		return res;
	}
}

public class Solution {
	private static boolean DEBUG = true;

	public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
		ListNode head = null;
		ListNode end = null;
		ListNode l = l1;
		ListNode r = l2;
		int carry = 0;

		while (l != null || r != null) {
			int t = 0;
			if (l != null) t += l.val;
			if (r != null) t += r.val;
			t += carry;

			ListNode temp = new ListNode(t % 10);
			carry = t / 10;
			if (head == null) {
				head = temp;
				end = temp;
			} else {
				end.next = temp;
				end = temp;
			}
			if (l != null) l = l.next;
			if (r != null) r = r.next;
		}

		if (carry != 0) end.next = new ListNode(carry);
		return head;
	}

	public static void main(String[] args) {
		Solution solution = new Solution();
		solution.judge(new int[] { 1 }, new int[] { 9, 9, 9, 9, 9, 9, 9, 9, 9, 9, 9 });
	}

	private void judge(int[] a1, int[] a2) {
		ListNode t = null, r = null;
		Solution solution = new Solution();

		t = new ListNode(a1);
		System.out.println(t.toString());
		r = new ListNode(a2);
		System.out.println(r.toString());

		ListNode ans = solution.addTwoNumbers(t, r);
		System.out.println(ans.toString());
	}
}